# **Data Loading and Insepection** #

In this step, we load the raw electricity and weather datasets and inspect their structure, size, data types, and completeness. This helps us understand the data distribution, identify missing values or inconsistenc
ies, and ensure the dataset is suitable for time-series analysis. Early inspection prevents downstream modeling errors and supports reliable anomaly detection.

### **Import Libraries** ###

In [ ]:
import pandas as pd
import numpy as np
import os



### **Data Loading** ###

In [27]:
print("Files in electricity raw folder:")
print(os.listdir())


Files in electricity raw folder:
['chilledwater.csv', 'electricity.csv', 'gas.csv', 'hotwater.csv', 'irrigation.csv', 'solar.csv', 'steam.csv', 'water.csv']


In [3]:
electricity_df = pd.read_csv(
    os.path.join(ELECTRICITY_PATH, "electricity.csv")
)

print("Electricity dataset loaded")
print("Shape:", electricity_df.shape)


Electricity dataset loaded
Shape: (17544, 1579)


In [28]:
electricity_df = pd.read_csv("electricity.csv")

In [ ]:

weather_df = pd.read_csv("weather.csv")

### **Inspection** ###

#### **Inspection on electricty dataset** ####

In [7]:
electricity_df.head()

timestamp  Panther_parking_Lorriane  Panther_lodging_Cora  \
0  2016-01-01 00:00:00                       0.0                   0.0   
1  2016-01-01 01:00:00                       0.0                   0.0   
2  2016-01-01 02:00:00                       0.0                   0.0   
3  2016-01-01 03:00:00                       0.0                   0.0   
4  2016-01-01 04:00:00                       0.0                   0.0   

   Panther_office_Hannah  Panther_lodging_Hattie  Panther_education_Teofila  \
0                    0.0                     0.0                        0.0   
1                    0.0                     0.0                        0.0   
2                    0.0                     0.0                        0.0   
3                    0.0                     0.0                        0.0   
4                    0.0                     0.0                        0.0   

   Panther_education_Jerome  Panther_retail_Felix  Panther_parking_Asia  \
0                       0.0                   0.0                   0.0   
1                       0.0                   0.0                   0.0   
2                       0.0                   0.0                   0.0   
3                       0.0                   0.0                   0.0   
4                       0.0                   0.0                   0.0   

   Panther_education_Misty  ...  Cockatoo_public_Caleb  \
0                      0.0  ...                123.200   
1                      0.0  ...                126.475   
2                      0.0  ...                127.825   
3                      0.0  ...                130.475   
4                      0.0  ...                129.675   

   Cockatoo_education_Tyler  Cockatoo_public_Shad  Mouse_health_Buddy  \
0                   727.575                69.200              8.8224   
1                   731.200                66.275             17.6449   
2                   724.675                64.675             17.6449   
3                   737.375                65.875             17.6449   
4                   721.150                66.275             17.6449   

   Mouse_health_Modesto  Mouse_lodging_Vicente  Mouse_health_Justin  \
0              370.0870                   10.0             282.9965   
1              737.8260                   30.0             574.9265   
2              729.9255                   30.0             570.2780   
3              722.2620                   20.0             561.1470   
4              719.1665                   30.0             564.3695   

   Mouse_health_Ileana  Mouse_health_Estela  Mouse_science_Micheal  
0                 26.0                135.0               168.2243  
1                 51.0                265.0               336.4486  
2                 50.0                272.0               336.4486  
3                 52.0                276.0               336.4486  
4                 50.0                280.0               336.4486  

[5 rows x 1579 columns]

In [8]:
electricity_df.tail()

timestamp  Panther_parking_Lorriane  Panther_lodging_Cora  \
17539  2017-12-31 19:00:00                   15.4830              135.2261   
17540  2017-12-31 20:00:00                   12.7224              135.9262   
17541  2017-12-31 21:00:00                   11.2822              135.1761   
17542  2017-12-31 22:00:00                   16.9233              137.6266   
17543  2017-12-31 23:00:00                   11.8223              136.1263   

       Panther_office_Hannah  Panther_lodging_Hattie  \
17539                 3.4357                 79.1353   
17540                 3.4087                 81.6958   
17541                 3.3546                 82.8160   
17542                 3.2876                 82.3359   
17543                 3.3686                 82.4959   

       Panther_education_Teofila  Panther_education_Jerome  \
17539                   105.6374                  465.0898   
17540                   107.7348                  463.6895   
17541                   106.1295                  461.2890   
17542                   109.6282                  460.5889   
17543                   103.7540                  462.0892   

       Panther_retail_Felix  Panther_parking_Asia  Panther_education_Misty  \
17539               67.0199               35.7069                  16.3231   
17540               56.6869               35.7069                  16.0831   
17541               55.0576               35.5068                  16.1631   
17542               49.6776               35.7069                  14.8829   
17543               42.6952               35.6069                  16.0031   

       ...  Cockatoo_public_Caleb  Cockatoo_education_Tyler  \
17539  ...                 96.925                   704.950   
17540  ...                 97.550                   695.700   
17541  ...                 93.825                   687.325   
17542  ...                 94.150                   674.275   
17543  ...                 96.325                   677.400   

       Cockatoo_public_Shad  Mouse_health_Buddy  Mouse_health_Modesto  \
17539               111.350                 0.0                   0.0   
17540               115.875                 0.0                   0.0   
17541               111.650                 0.0                   0.0   
17542               111.950                 0.0                   0.0   
17543               113.250                 0.0                   0.0   

       Mouse_lodging_Vicente  Mouse_health_Justin  Mouse_health_Ileana  \
17539                    0.0                  0.0                  0.0   
17540                    0.0                  0.0                  0.0   
17541                    0.0                  0.0                  0.0   
17542                    0.0                  0.0                  0.0   
17543                    0.0                  0.0                  0.0   

       Mouse_health_Estela  Mouse_science_Micheal  
17539                  0.0                    0.0  
17540                  0.0                    0.0  
17541                  0.0                    0.0  
17542                  0.0                    0.0  
17543                  0.0                    0.0  

[5 rows x 1579 columns]

#### **Dataset Shape** ####

In [9]:
print("Electricity data shape:", electricity_df.shape)

Electricity data shape: (17544, 1579)


#### **Columns** ####

In [10]:
electricity_df.columns

Index(['timestamp', 'Panther_parking_Lorriane', 'Panther_lodging_Cora',
       'Panther_office_Hannah', 'Panther_lodging_Hattie',
       'Panther_education_Teofila', 'Panther_education_Jerome',
       'Panther_retail_Felix', 'Panther_parking_Asia',
       'Panther_education_Misty',
       ...
       'Cockatoo_public_Caleb', 'Cockatoo_education_Tyler',
       'Cockatoo_public_Shad', 'Mouse_health_Buddy', 'Mouse_health_Modesto',
       'Mouse_lodging_Vicente', 'Mouse_health_Justin', 'Mouse_health_Ileana',
       'Mouse_health_Estela', 'Mouse_science_Micheal'],
      dtype='object', length=1579)

#### **Missing Value** ####

In [11]:
electricity_df.isna().sum()

timestamp                     0
Panther_parking_Lorriane     11
Panther_lodging_Cora         11
Panther_office_Hannah        12
Panther_lodging_Hattie       12
                           ... 
Mouse_lodging_Vicente         0
Mouse_health_Justin           0
Mouse_health_Ileana         165
Mouse_health_Estela           0
Mouse_science_Micheal         0
Length: 1579, dtype: int64

##### **Missing Value (%)** #####

In [12]:
(electricity_df.isna().mean() * 100).sort_values(ascending=False)

Eagle_lodging_Garland      100.000000
Rat_public_Ulysses         100.000000
Bobcat_education_Barbra     99.373005
Bobcat_education_Seth       96.243730
Rat_education_Mac           94.795942
                              ...    
Lamb_public_Angeline         0.000000
Lamb_education_Harold        0.000000
Lamb_office_Peggy            0.000000
Lamb_office_William          0.000000
Eagle_assembly_Ian           0.000000
Length: 1579, dtype: float64

##### **Statistical Summary** #####

In [13]:
electricity_df.describe()

Panther_parking_Lorriane  Panther_lodging_Cora  Panther_office_Hannah  \
count              17533.000000          17533.000000           17532.000000   
mean                   8.661108            109.326952               5.396102   
std                    4.900062             56.498653               4.223728   
min                    0.000000              0.000000               0.000000   
25%                    6.661300            108.270900               2.676500   
50%                    9.601900            126.624400               4.761900   
75%                   12.662500            145.278000               7.952000   
max                   25.497200            285.451200              27.770400   

       Panther_lodging_Hattie  Panther_education_Teofila  \
count            17532.000000               17533.000000   
mean               113.520501                 126.329541   
std                 61.636034                  71.783004   
min                  0.000000                   0.000000   
25%                 89.937400                 100.849500   
50%                133.705800                 141.866400   
75%                158.270500                 179.404600   
max                298.053200                 375.038400   

       Panther_education_Jerome  Panther_retail_Felix  Panther_parking_Asia  \
count              17534.000000          17533.000000          17533.000000   
mean                 392.126185             96.513085             21.725700   
std                  195.116299             62.818199             13.532895   
min                    0.000000              0.000000              0.000000   
25%                  444.385800             56.097800             16.003100   
50%                  469.390600             96.100600             18.003500   
75%                  498.371200            148.136600             35.806900   
max                 1052.711000            279.125900             54.804300   

       Panther_education_Misty  Panther_retail_Gilbert  ...  \
count             17524.000000            17532.000000  ...   
mean                 27.257845                0.594539  ...   
std                  15.092654                0.408065  ...   
min                   0.000000                0.000000  ...   
25%                  17.923500                0.330100  ...   
50%                  34.646700                0.550100  ...   
75%                  39.047500                0.890200  ...   
max                  74.733100                1.520300  ...   

       Cockatoo_public_Caleb  Cockatoo_education_Tyler  Cockatoo_public_Shad  \
count           16110.000000              16116.000000          16113.000000   
mean              181.787028                770.348807            129.774161   
std                54.150864                 47.807854             37.264065   
min                83.350000                548.800000             60.050000   
25%               133.425000                734.568750             97.525000   
50%               172.487500                763.925000            126.550000   
75%               228.875000                803.293750            158.825000   
max               316.350000                925.550000            278.575000   

       Mouse_health_Buddy  Mouse_health_Modesto  Mouse_lodging_Vicente  \
count        17483.000000          17540.000000           17544.000000   
mean             6.134131            482.296901              41.447504   
std              8.463139            266.172973              18.780824   
min              0.000000              0.000000               0.000000   
25%              0.000000            310.030500              30.000000   
50%              0.000000            433.645000              40.000000   
75%             17.644900            671.243250              60.000000   
max             40.816300           1267.846000             100.000000   

       Mouse_health_Justin  Mouse_health_Ileana  Mouse_health_Estela  \
count

##### **Structure** #####

In [14]:
electricity_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17544 entries, 0 to 17543
Columns: 1579 entries, timestamp to Mouse_science_Micheal
dtypes: float64(1578), object(1)
memory usage: 211.3+ MB


##### **Data Types** #####

In [15]:
electricity_df.dtypes

timestamp                    object
Panther_parking_Lorriane    float64
Panther_lodging_Cora        float64
Panther_office_Hannah       float64
Panther_lodging_Hattie      float64
                             ...   
Mouse_lodging_Vicente       float64
Mouse_health_Justin         float64
Mouse_health_Ileana         float64
Mouse_health_Estela         float64
Mouse_science_Micheal       float64
Length: 1579, dtype: object

### **Key findings** ###
- Rows: 17,544 hourly timestamps starting from (2016-01-01 00:00:00) till (2017-12-31 23:00:00)
- Columns: 1579
  - 1st = timestamp (object)
  - others = 1578 building/meter energy readings (float64)
- Memory Usage: ~ 211 MB (very large)
- Missing Value:
  - no missing value in timestamp col
  - multiple in meter columns
- Timestamp is not yet datetime (need to be converted)

#### **Inspection on weather dataset** ####

In [16]:
weather_df.head()

timestamp  site_id  airTemperature  cloudCoverage  \
0  2016-01-01 00:00:00  Panther            19.4            NaN   
1  2016-01-01 01:00:00  Panther            21.1            6.0   
2  2016-01-01 02:00:00  Panther            21.1            NaN   
3  2016-01-01 03:00:00  Panther            20.6            NaN   
4  2016-01-01 04:00:00  Panther            21.1            NaN   

   dewTemperature  precipDepth1HR  precipDepth6HR  seaLvlPressure  \
0            19.4             0.0             NaN             NaN   
1            21.1            -1.0             NaN          1019.4   
2            21.1             0.0             NaN          1018.8   
3            20.0             0.0             NaN          1018.1   
4            20.6             0.0             NaN          1019.0   

   windDirection  windSpeed  
0            0.0        0.0  
1            0.0        0.0  
2          210.0        1.5  
3            0.0        0.0  
4          290.0        1.5

In [17]:
weather_df.tail()

timestamp site_id  airTemperature  cloudCoverage  \
331161  2017-12-31 19:00:00   Mouse             8.5            NaN   
331162  2017-12-31 20:00:00   Mouse             8.5            NaN   
331163  2017-12-31 21:00:00   Mouse             8.2            NaN   
331164  2017-12-31 22:00:00   Mouse             7.5            NaN   
331165  2017-12-31 23:00:00   Mouse             7.2            NaN   

        dewTemperature  precipDepth1HR  precipDepth6HR  seaLvlPressure  \
331161             4.8             NaN             NaN           992.3   
331162             4.5             NaN             NaN           992.1   
331163             4.0             NaN             NaN           992.1   
331164             4.3             NaN             NaN           993.7   
331165             3.7             NaN             NaN           995.7   

        windDirection  windSpeed  
331161          210.0        8.2  
331162          210.0        7.2  
331163          230.0       10.3  
331164          260.0       12.9  
331165          260.0       10.3

##### **Dataset Shape** #####

In [18]:
print("Weather data shape:", weather_df.shape)

Weather data shape: (331166, 10)


##### **Columns** #####

In [19]:
weather_df.columns

Index(['timestamp', 'site_id', 'airTemperature', 'cloudCoverage',
       'dewTemperature', 'precipDepth1HR', 'precipDepth6HR', 'seaLvlPressure',
       'windDirection', 'windSpeed'],
      dtype='object')

##### **Missing Value** #####

In [20]:
weather_df.isna().sum()

timestamp              0
site_id                0
airTemperature       128
cloudCoverage     170987
dewTemperature       328
precipDepth1HR    133186
precipDepth6HR    313004
seaLvlPressure     21624
windDirection      13005
windSpeed            574
dtype: int64

##### **Missing Value(%)** #####

In [21]:
(weather_df.isna().mean() * 100).sort_values(ascending=False)

precipDepth6HR    94.515741
cloudCoverage     51.631810
precipDepth1HR    40.217293
seaLvlPressure     6.529656
windDirection      3.927034
windSpeed          0.173327
dewTemperature     0.099044
airTemperature     0.038651
site_id            0.000000
timestamp          0.000000
dtype: float64

##### **Statistical summary** #####

In [22]:
weather_df.describe()

airTemperature  cloudCoverage  dewTemperature  precipDepth1HR  \
count   331038.000000  160179.000000   330838.000000   197980.000000   
mean        14.235343       1.920907        7.649370        0.955738   
std          9.990392       2.550744        9.201438        8.273852   
min        -28.900000       0.000000      -35.000000       -1.000000   
25%          7.800000       0.000000        1.800000        0.000000   
50%         14.400000       0.000000        8.500000        0.000000   
75%         21.100000       4.000000       13.900000        0.000000   
max         48.300000       9.000000       26.700000      597.000000   

       precipDepth6HR  seaLvlPressure  windDirection      windSpeed  
count    18162.000000   309542.000000  318161.000000  330592.000000  
mean        13.536560     1016.063498     184.391299       3.569554  
std         43.801017        8.052463     111.571354       2.335197  
min         -1.000000      968.200000       0.000000       0.000000  
25%          0.000000     1011.600000      90.000000       2.100000  
50%          0.000000     1016.200000     200.000000       3.100000  
75%          5.000000     1020.900000     280.000000       5.000000  
max        770.000000     1050.100000     360.000000      24.200000

##### **Structure** #####

In [23]:
weather_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 331166 entries, 0 to 331165
Data columns (total 10 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   timestamp       331166 non-null  object 
 1   site_id         331166 non-null  object 
 2   airTemperature  331038 non-null  float64
 3   cloudCoverage   160179 non-null  float64
 4   dewTemperature  330838 non-null  float64
 5   precipDepth1HR  197980 non-null  float64
 6   precipDepth6HR  18162 non-null   float64
 7   seaLvlPressure  309542 non-null  float64
 8   windDirection   318161 non-null  float64
 9   windSpeed       330592 non-null  float64
dtypes: float64(8), object(2)
memory usage: 25.3+ MB


##### **Data types** #####

In [24]:
weather_df.dtypes

timestamp          object
site_id            object
airTemperature    float64
cloudCoverage     float64
dewTemperature    float64
precipDepth1HR    float64
precipDepth6HR    float64
seaLvlPressure    float64
windDirection     float64
windSpeed         float64
dtype: object

### **Key Findings:** ###

- Rows: 331,166 hourly timestamps starting from (2016-01-01 00:00:00) till (2017-12-31 23:00:00)
- Columns: 10
  - timestamp (object)
  - site_d (object)
  - other key feature (float64)
- Key Features:
  - airTemperature
  - dewTemperature
  - windSpeed
  - seaLvlPressure
  - cloudCoverage
  - precipDepth1HR
  - precipDepth6HR
- Timestamp: Object (will need conversion later)
- Missing Values:
  - High missing values in:
    - precipDepth6HR
    - cloudCoverage
- Memory Usage: ~25 MB (manageable)

## **Observation** ##

#### Electricity Dataset
- The electricity dataset contains energy consumption readings for 1,578 buildings/meters that spans approximately two years of hourly data.
- The dataset is wide, with each column representing a separate building or meter.
- The timestamp column is currently of type object and will need conversion to datetime.
- A significant number of missing values are present across different meter columns likely due to sensor downtime.
- The large size and multivariate nature of the dataset make it suitable for time-series anomaly detection across multiple entities.

#### Weather Dataset
- The weather dataset provides contextual environmental information such as temperature, wind speed, pressure, and precipitation.
- It contains multiple weather variables recorded over timestamp.
- Some features (e.g., precipDepth6HR, cloudCoverage) contain a high percentage of missing values.
- The dataset is relatively structured and consistent, making it suitable for use as external contextual features during anomaly detection.

#### Overall Assessment
- Both datasets are aligned temporally and are appropriate for time-series energy anomaly detection.
- Electricity data serves as the core signal, while weather data provides valuable contextual signals.
- No data cleaning or transformations have been applied at this stage, as this notebook is strictly focused on data understanding.